### PR TITLE
feat(storage): rename new StateStore to LocalStateStore and add new_local method

### PR DESCRIPTION
**Pull request by wenym1**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

In this PR, we turn the newly added `StateStore` trait into `LocalStateStore`, which is dedicated to serve streaming read and write, while the original `StateStore` will be preserved for serving batch query. We added a new associated type named `Local`  in `StateStore` with trait bound `LocalStateStore`, and added a new method `new_local`, which instantiates a new local state store instance.

For better reuse of code and for clearer trait functionality structure, we split the current `StateStore` into `StateStoreRead`, `StateStoreScan` and `StateStoreWrite`. The `StateStore` will be the combination all of them, while the `LocalStateStore` only implements `StateStoreRead` and `StateStoreWrite` without `Clone`. In the future if we want to make `StateStore` read only, we can simply remove the `StateStoreWrite` from it.

Besides, the current two `StateStore` have different read(get, iter) interface. In this PR, since we introduce a common trait `StateStoreRead`, we unify the two read interfaces to the newer one. `backward_iter` and `backward_scan` is removed until they will be supported.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

### CODE REVIEW

**Review by Li0k — APPROVED** [review]
Rest LGTM. PTAL @hzxa21

File: src/storage/src/table/streaming_table/state_table.rs [hunk 1]
```diff
@@ -910,8 +920,9 @@ impl<S: StateStore> StateTable<S> {
     }
 }
 
-pub type RowStream<'a, S: StateStore> = impl Stream<Item = StorageResult<Cow<'a, Row>>>;
-pub type RowStreamWithPk<'a, S: StateStore> =
+pub type RowStream<'a, S: StateStoreRead + StateStoreWrite> =
+    impl Stream<Item = StorageResult<Cow<'a, Row>>>;
```
Comment (wcy-fdu) [inline src/storage/src/table/streaming_table/state_table.rs:924]: `RowStream ` and `StorageIterInner ` may only need `StateStoreRead `.

**Review by wcy-fdu — COMMENTED** [review]
Will `StorageTable` use `StateStoreRead ` in later PR?

**Comment by wcy-fdu** [comment]
@st1page  @BugenZhao  PTAL😊

File: src/storage/src/memory.rs [hunk 2]
```diff
@@ -283,28 +292,16 @@ impl StateStore for MemoryStateStore {
...
 
-    fn backward_iter(
-        &self,
-        _key_range: (Bound<Vec<u8>>, Bound<Vec<u8>>),
-        _read_options: ReadOptions,
-    ) -> Self::BackwardIterFuture<'_> {
-        async move { unimplemented!() }
-    }
+impl StateStore for MemoryStateStore {
+    type Local = Self;
```
Comment (BugenZhao) [inline src/storage/src/memory.rs:300]: 🚀
```suggestion
    type Local = Self;
```

File: src/storage/src/hummock/state_store_v1.rs [hunk 3]
```diff
@@ -636,6 +616,10 @@ impl StateStore for HummockStorageV1 {
             Ok(())
         }
     }
+
+    fn new_local(&self, _table_id: TableId) -> Self::NewLocalFuture<'_> {
+        async { self.clone() }
```
Comment (wcy-fdu) [inline src/storage/src/hummock/state_store_v1.rs:621]: May cause executors' new to async, is this by design?cc @BugenZhao

File: src/storage/src/store.rs [hunk 4]
```diff
@@ -172,37 +227,64 @@ pub trait StateStore: Send + Sync + 'static + Clone {
...
-    type Item;
-    type NextFuture<'a>: Future<Output = StorageResult<Option<Self::Item>>> + Send;
+/// A state store that is dedicated for streaming operator, which only reads the uncommitted data
+/// written by itself. Each local state store is not `Clone`, and is owned by a streaming state
+/// table.
+pub trait LocalStateStore: StateStoreRead + StateStoreWrite + StaticSendSync {
+    /// Inserts a key-value entry associated with a given `epoch` into the state store.
+    fn insert(&self, _key: Bytes, _val: Bytes) -> StorageResult<()> {
+        unimplemented!()
+    }
```
Comment (BugenZhao) [inline src/storage/src/store.rs:241]: Can we remove these default implementations?

**Review by BugenZhao — APPROVED** [review]

**Review by wcy-fdu — APPROVED** [review]
LGTM to the `StateTable` part, I think we can merge this PR first, and we may need to discuss whether changing executors's `new` to async the next PR.

File: src/storage/src/store.rs (hunk 4 above)
Comment (wenym1) [reply src/storage/src/store.rs:241]: These methods are not used yet, but we may use them later. If we remove the default implementation, we will have to add the `unimplemented` in all the types that implement `LocalStateStore`. So I will prefer keeping the default implementation.

**Pull request merged**